moto/cloudformation/custom_model.py

```diff
@@ -0,0 +1,112 @@
+import threading
+
+from moto import settings
+from moto.core.models import CloudFormationModel
+from moto.awslambda import lambda_backends
+from uuid import uuid4
+
+
+class CustomModel(CloudFormationModel):
+    def __init__(self, region_name, request_id, logical_id, resource_name):
+        self.region_name = region_name
+        self.request_id = request_id
+        self.logical_id = logical_id
+        self.resource_name = resource_name
+        self.data = dict()
+        self._finished = False
+
+    def set_data(self, data):
+        self.data = data
+        self._finished = True
+
+    def is_created(self):
+        return self._finished
+
+    @property
+    def physical_resource_id(self):
+        return self.resource_name
+
+    @staticmethod
+    def cloudformation_type():
+        return "?"
+
+    @classmethod
+    def create_from_cloudformation_json(
+        cls, resource_name, cloudformation_json, region_name, **kwargs
+    ):
+        print("create_from_cloudformation")
+        logical_id = kwargs["LogicalId"]
+        stack_id = kwargs["StackId"]
+        resource_type = kwargs["ResourceType"]
+        properties = cloudformation_json["Properties"]
+        service_token = properties["ServiceToken"]
+
+        backend = lambda_backends[region_name]
+        fn = backend.get_function(service_token)
+
+        request_id = str(uuid4())
+
+        custom_resource = CustomModel(
+            region_name, request_id, logical_id, resource_name
+        )
+
+        from moto.cloudformation import cloudformation_backends
+
+        stack = cloudformation_backends[region_name].get_stack(stack_id)
+        stack.add_custom_resource(custom_resource)
+
+        event = {
+            "RequestType": "Create",
+            "ServiceToken": service_token,
+            # A request will be send to this URL to indicate success/failure
+            # This request will be coming from inside a Docker container
+            # Note that, in order to reach the Moto host, the Moto-server should be listening on 0.0.0.0
+            #
+            # Alternative: Maybe we should let the user pass in a container-name where Moto is running?
+            # Similar to how we know for sure that the container in our CI is called 'motoserver'
+            "ResponseURL": f"{settings.moto_server_host()}/cloudformation_{region_name}/cfnresponse?stack={stack_id}",
+            "StackId": stack_id,
+            "RequestId": request_id,
+            "LogicalResourceId": logical_id,
+            "ResourceType": resource_type,
+            "ResourceProperties": properties,
+        }
+        print(event)
+
+        def invoke_function():
+            try:
+                fn.invoke(json.dumps(event), {}, {})
+            except Exception as e:
+                print(e)
+
+        print("Invoking function")
+        # p = multiprocessing.Process(target=invoke_function)
+        # p.start()
+        invoke_thread = threading.Thread(target=invoke_function, args=())
```

Comment:
The custom Lambda needs to be invoked async, as we're still in the `create_stack`-call, and we don't want to wait 10+ seconds for Docker to be invoked